src/wp-includes/interactivity-api/class-wp-interactivity-api.php

```diff
@@ -167,10 +167,54 @@ public function print_client_interactivity_data() {
 		}
 
 		if ( ! empty( $interactivity_data ) ) {
+
+			/*
+			 * This data will be printed as JSON inside a script tag like this:
+			 *   <script type="application/json"></script>
+			 *
+			 * A script tag must be closed by a sequence beginning with `</`. It's impossible to
+			 * close a script tag without using `<`. We ensure that `<` is escaped and `/` can
+			 * remain unescaped, so `</script>` will be printed as `\u003C/script\u00E3`.
+			 *
+			 *   - JSON_HEX_TAG: All < and > are converted to \u003C and \u003E.
+			 *   - JSON_UNESCAPED_SLASHES: Don't escape /.
+			 *
+			 * @see https://www.php.net/manual/en/json.constants.php for details on these constants.
+			 * @see https://html.spec.whatwg.org/#script-data-state for details on script
+			 *      tag parsing.
+			 */
+			$json_encode_flags = JSON_HEX_TAG | JSON_UNESCAPED_SLASHES;
+			if ( 'UTF-8' === get_option( 'blog_charset' ) ) {
+				/*
+				 * If the page will use UTF-8 encoding, it's safe to print unescaped unicode in
+				 * JSON. Set the following flags:
+				 *
+				 * - JSON_UNESCAPED_UNICODE: Encode multibyte Unicode characters literally
+				 *   (default is to escape as \uXXXX).
+				 * - JSON_UNESCAPED_LINE_TERMINATORS: The line terminators are kept unescaped when
```

Comment:
Why not escape line terminators? It seems like my normal expectation is that they be escaped.

Reply:
It's a poorly named and poorly documented flag, but [it causes `\u2028` (LINE SEPARATOR) and `\u2029` (PARAGRAPH SEPARATOR) _not_ to be escaped](https://github.com/php/php-src/blob/0dfd2a93ab70e1a6e09f3fad2e22f50990da6f9d/ext/json/json_encoder.c#L399-L405). This is included in the updated tests.

It has nothing to do with newlines or carriage returns or line feeds or anything like you might expect to be a "line terminator" 🤷. These two characters are not in the restricted JSON characters so its fine to print them.

Reply:
That may be a good point, `\u2028` is easier to spot than whitespace. At the same time, we won't be encoding `\u00a0` or most of the other whitespace characters that are likely much more common.

I'd also argue that the only place we control this is where they're printed in the HTML. Assuming that _something_ reads the string out of HTML and parses the JSON —probably JavaScript with `JSON.parse`—, it doesn't matter how it was encoded when serialized.

I have a slight preference to keep this flag and leave these unescaped, but I don't feel strongly one way or the other.

Reply:
Here's what it looks like, this is `"before\u2028middle\u2029after"`:

| Browser | Elements panel | view source |
|--------|--------|--------|
| Chrome | ![Screenshot 2024-05-10 at 21 01 48](https://github.com/WordPress/wordpress-develop/assets/841763/ef2019c6-c9c7-40f6-8144-9bc26af5c3e1) | ![Screenshot 2024-05-10 at 20 34 35](https://github.com/WordPress/wordpress-develop/assets/841763/6e728005-1590-4534-9311-2a0c6e0c798b) |
| Firefox | ![Screenshot 2024-05-10 at 21 03 02](https://github.com/WordPress/wordpress-develop/assets/841763/23e80857-98a0-415f-9dc6-68ace886bf4b) | ![Screenshot 2024-05-10 at 21 05 33](https://github.com/WordPress/wordpress-develop/assets/841763/c70e98f8-c5dd-4e7f-b173-eb031c11e0d8) |

Safari looks like Firefox, the characters are invisible and it reads `"beforemiddleafter"`.